File: src/test/java/org/wikimedia/diffdb/TestSearcherDaemon.java
package org.wikimedia.diffdb;
import org.junit.*;
import java.io.*;
import java.util.*;
import java.net.*;
import org.apache.lucene.index.*;
import org.apache.lucene.store.*;
import org.apache.lucene.search.*;
import org.apache.lucene.document.*;
import org.apache.lucene.queryParser.QueryParser;
import org.apache.lucene.util.Version;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import static org.junit.Assert.*;

public class TestSearcherDaemon {
  private static File newTempFile(String content) throws IOException {
    File file = File.createTempFile("indexer", ".txt");
    file.deleteOnExit();
    BufferedWriter writer = new BufferedWriter(new FileWriter(file));
    writer.write(content);
    writer.close();
    return file;
  }

  private static String readAll(Reader reader) throws IOException {
    StringBuffer buff = new StringBuffer();
    char[] buffb = new char[4096];
    int len;
    while ( (len = reader.read(buffb)) > 0 ) {
      buff.append(buffb, 0, len);
    }
    return buff.toString();
  }


  private static JSONObject retrieve(InetSocketAddress address, JSONObject query) throws IOException, JSONException {
    Socket socket = new Socket(address.getAddress(), address.getPort());
    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()));
    writer.write(query.toString());
    writer.flush();
    socket.shutdownOutput();
    BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
    JSONObject ret = new JSONObject(readAll(reader));
    writer.close();
    reader.close();
    socket.close();
    return ret;
  }

  private static InetSocketAddress findFreeAddress() throws IOException {
    ServerSocket socket = new ServerSocket(0);
    InetSocketAddress address = new InetSocketAddress(socket.getLocalPort());
    socket.close();
    return address;
  }

  @Test public void smallDocuments() throws IOException, JSONException, InterruptedException {
    Directory dir = new RAMDirectory();
    IndexWriter writer = new IndexWriter(dir,
                                         new IndexWriterConfig(Version.LUCENE_35,
                                                               new SimpleNGramAnalyzer(3)));
    Indexer indexer = new Indexer(writer, 2, 2, 100);
    indexer.indexDocuments(newTempFile("233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'\n" +
                                       "18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'\n"));
    indexer.finish();

    IndexSearcher searcher = new IndexSearcher(IndexReader.open(dir));
    InetSocketAddress address = findFreeAddress();
    new Thread(new SearcherDaemon(address, searcher, new QueryParser(Version.LUCENE_35, "added", new SimpleNGramAnalyzer(3)))).start();

    Thread.sleep(1000L);

    // query "/Todo" and receive rev_ids
    {
      JSONObject q = new JSONObject();
      q.put("q", "/Todo");
      q.put("fields", "rev_id");
      JSONObject json = retrieve(address, q);
      System.err.println(json);//!
      assertEquals(1, json.getInt("hits_all"));
      assertEquals(18201, json.getJSONArray("hits").getJSONArray(0).getInt(0));
    }
    // query "Accessible" and receive rev_ids and timestamps
    {
      JSONObject q = new JSONObject();
      q.put("q", "Accessible");
      q.put("fields", new JSONArray(new String[]{"rev_id", "timestamp"}));
      JSONObject json = retrieve(address, q);
      assertEquals(1, json.getInt("hits_all"));
      assertEquals(233192,
                   json.getJSONArray("hits").getJSONArray(0).getInt(0));
      assertEquals("2001-01-21T11:12:21Z",
                   json.getJSONArray("hits").getJSONArray(0).getString(1));
    }
  }

  @Test public void smallDocumentsComplexQueries() throws IOException, JSONException, InterruptedException {
    Directory dir = new RAMDirectory();
    IndexWriter writer = new IndexWriter(dir,
                                         new IndexWriterConfig(Version.LUCENE_35,
                                                               new SimpleNGramAnalyzer(3)));
    Indexer indexer = new Indexer(writer, 2, 2, 100);
    indexer.indexDocuments(newTempFile("233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'\n" +
                                       "18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'\n"));
    indexer.finish();

    IndexSearcher searcher = new IndexSearcher(IndexReader.open(dir));
    InetSocketAddress address = findFreeAddress();
    new Thread(new SearcherDaemon(address, searcher, new QueryParser(Version.LUCENE_35, "added", new SimpleNGramAnalyzer(3)))).start();

    Thread.sleep(1000L);

    // query namespace:0 and page_id:12 and receive rev_ids
    {
      JSONObject q = new JSONObject();
      q.put("q", "namespace:0 page_id:12");
      q.put("fields", "rev_id");
      JSONObject json = retrieve(address, q);
      System.err.println(json);//!
      assertEquals(1, json.getInt("hits_all"));
      assertEquals(18201, json.getJSONArray("hits").getJSONArray(0).getInt(0));
    }
  }

  @Test public void smallDocumentsWithCollapsedHits() throws IOException, JSONException, InterruptedException {
    Directory dir = new RAMDirectory();
    IndexWriter writer = new IndexWriter(dir,
                                         new IndexWriterConfig(Version.LUCENE_35,
                                                               new SimpleNGramAnalyzer(1)));
    Indexer indexer = new Indexer(writer, 2, 10, 100);
    indexer.indexDocuments(newTempFile("233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'\n" +
                                       "18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'\n" +
                                       "18210	12	0	'Anarchism'	1014749333	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'\n"));
    indexer.finish();

    IndexSearcher searcher = new IndexSearcher(IndexReader.open(dir));
    InetSocketAddress address = findFreeAddress();
    new Thread(new SearcherDaemon(address, searcher, new QueryParser(Version.LUCENE_35, "added", new SimpleNGramAnalyzer(1)))).start();

    Thread.sleep(1000L);

    // query "/Todo" and receive rev_ids monthly
    {
      JSONObject q = new JSONObject();
      q.put("q", "/Todo");
      q.put("fields", "rev_id");
      q.put("collapse_hits", "month");
      JSONObject json = retrieve(address, q);
      System.err.println(json);//!
      assertEquals(2, json.getInt("hits_all"));
      assertEquals("2002-02", json.getJSONArray("hits").getJSONArray(0).getString(0));
      assertEquals(18201,     json.getJSONArray("hits").getJSONArray(0).getJSONArray(1).getJSONArray(0).getInt(0));
      assertEquals(18210,     json.getJSONArray("hits").getJSONArray(0).getJSONArray(1).getJSONArray(1).getInt(0));
    }
    // query "/Todo" and receive rev_ids daily
    {
      JSONObject q = new JSONObject();
      q.put("q", "/Todo");
      q.put("fields", "rev_id");
      q.put("collapse_hits", "day");
      JSONObject json = retrieve(address, q);
      assertEquals(2, json.getInt("hits_all"));
      System.err.println(json);//!
      assertEquals("2002-02-26", json.getJSONArray("hits").getJSONArray(0).getString(0));
      assertEquals("2002-02-27", json.getJSONArray("hits").getJSONArray(1).getString(0));
      assertEquals(18201,        json.getJSONArray("hits").getJSONArray(0).getJSONArray(1).getJSONArray(0).getInt(0));
      assertEquals(18210,        json.getJSONArray("hits").getJSONArray(1).getJSONArray(1).getJSONArray(0).getInt(0));
    }
    // query "/Todo" and receive daily counts
    {
      JSONObject q = new JSONObject();
      q.put("q", "/Todo");
      q.put("collapse_hits", "day");
      JSONObject json = retrieve(address, q);
      assertEquals(2, json.getInt("hits_all"));
      System.err.println(json);//!
      assertEquals("2002-02-26", json.getJSONArray("hits").getJSONArray(0).getString(0));
      assertEquals("2002-02-27", json.getJSONArray("hits").getJSONArray(1).getString(0));
      assertEquals(1,            json.getJSONArray("hits").getJSONArray(0).getInt(1));
      assertEquals(1,            json.getJSONArray("hits").getJSONArray(1).getInt(1));
    }
  }

  @Test public void smallDocumentsHash() throws IOException, JSONException, InterruptedException {
    Directory dir = new RAMDirectory();
    IndexWriter writer = new IndexWriter(dir,
                                         new IndexWriterConfig(Version.LUCENE_35,
                                                               new HashedNGramAnalyzer(3, 3, 11)));
    Indexer indexer = new Indexer(writer, 2, 2, 100);
    indexer.indexDocuments(newTempFile("233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'\n" +
                                       "18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'\n"));
    indexer.finish();

    IndexSearcher searcher = new IndexSearcher(IndexReader.open(dir));
    InetSocketAddress address = findFreeAddress();
    new Thread(new SearcherDaemon(address, searcher, new QueryParser(Version.LUCENE_35, "added", new HashedNGramAnalyzer(3, 3, 11)))).start();

    Thread.sleep(1000L);

    // query "/Todo" and receive rev_ids
    {
      JSONObject q = new JSONObject();
      q.put("q", "/Todo");
      q.put("fields", "rev_id");
      JSONObject json = retrieve(address, q);
      System.err.println(json);//!
      assertEquals(1, json.getInt("hits_all"));
      assertEquals(18201, json.getJSONArray("hits").getJSONArray(0).getInt(0));
    }
    // query "Accessible" and receive rev_ids and timestamps
    {
      JSONObject q = new JSONObject();
      q.put("q", "Accessible");
      q.put("fields", new JSONArray(new String[]{"rev_id", "timestamp"}));
      JSONObject json = retrieve(address, q);
      assertEquals(1, json.getInt("hits_all"));
      assertEquals(233192,
                   json.getJSONArray("hits").getJSONArray(0).getInt(0));
      assertEquals("2001-01-21T11:12:21Z",
                   json.getJSONArray("hits").getJSONArray(0).getString(1));
    }
  }
}

/*
 * Local variables:
 * tab-width: 2
 * c-basic-offset: 2
 * indent-tabs-mode: nil
 * End:
 */

File: src/test/scala/test-searcherdaemon.scala
import org.scalatest.junit.JUnitRunner
import org.junit.runner.RunWith
import org.scalatest.FunSpec
import scala.collection.JavaConversions._

import org.wikimedia.revdiffsearch._
import java.io._
import java.util.logging._
import java.net._
import org.apache.lucene.index._
import org.apache.lucene.analysis._
import org.apache.lucene.store._
import org.apache.lucene.search._
import org.apache.lucene.document._
import org.apache.lucene.queryparser.classic.QueryParser
import org.apache.lucene.util.Version
import scala.io
import org.json4s.{JInt, JString, JField, JArray, JValue, JObject}
import org.json4s.native.{JsonMethods => JM}
import org.json4s.JsonDSL._
import scala.language.reflectiveCalls

@RunWith(classOf[JUnitRunner])
class SearcherDaemonSpec extends FunSpec {

  def newTempFile(content: String): File = {
    val file = File.createTempFile("indexer", ".txt")
    file.deleteOnExit
    val writer = new BufferedWriter(new FileWriter(file))
    writer.write(content)
    writer.close()
    file
  }

  def retrieve(address: InetSocketAddress, query: JValue): JValue = {
    val socket = new Socket(address.getAddress, address.getPort)
    val writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()))
    writer.write(JM.compact(JM.render(query)))
    writer.write("\n")
    writer.flush()
    socket.shutdownOutput()
    val ret = JM.parse(io.Source.fromInputStream(socket.getInputStream).mkString)
    writer.close
    socket.close
    info(JM.compact(JM.render(ret)))
    ret
  }

  def findFreeAddress(): InetSocketAddress = {
    val socket = new ServerSocket(0)
    val address = new InetSocketAddress(socket.getLocalPort())
    socket.close
    address
  }

  def waitUntilPrepared(address: InetSocketAddress, limit: Long): Boolean = {
    val start = System.currentTimeMillis
    while ( true) {
      try {
        val sock = new Socket(address.getAddress(), address.getPort())
        if ( sock.isConnected ) {
          sock.close
          return true
        }
      } catch {
        case e: ConnectException => {
          Thread.sleep(limit / 20)
          if ( System.currentTimeMillis - start > limit ) {
            return false
          }
        }
      }
    }
    false
  }

  Logger.getLogger(classOf[SearcherDaemon].getName()).setLevel(Level.INFO);
  Logger.getLogger(classOf[Indexer].getName()).setLevel(Level.WARNING);


  def daemon(analyzer: Analyzer, doc: String, http: Boolean=false) = {
    new {
      val dir = new RAMDirectory()
      val writer = new IndexWriter(dir,
                                   new IndexWriterConfig(Version.LUCENE_44,
                                                         analyzer))
      val indexer = new Indexer(writer, 2, 2, 100)
      indexer.indexDocuments(newTempFile(doc))
      indexer.finish
      
      val searcher = new IndexSearcher(DirectoryReader.open(dir))
      val address = findFreeAddress()
      new Thread(new SearcherDaemon(address, searcher, new QueryParser(Version.LUCENE_44, "added", RevDiffSearchUtils.getAnalyzerCombined(analyzer)), 1, http)).start()
      waitUntilPrepared(address, 1000L)
    }
  }

  def n3daemon(doc: String) = daemon(new SimpleNGramAnalyzer(3), doc)
  def n1daemon(doc: String) = daemon(new SimpleNGramAnalyzer(1), doc)


  describe("SearcherDaemon with two small documents") {

    val doc = """233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'
18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'This covers subject to'	9304:-1:u'/Talk'	9464:1:u'\\n'
"""
    val d = n3daemon(doc)
      
    it("should process simple string queries to get rev_ids") {
      val q = ("q" -> "\"/Todo\"")~("fields" -> "rev_id")
      val json = retrieve(d.address, q)
      assert(JInt(1) == (json \ "hits_all"))
      assert(JString("18201") == (json \ "hits")(0)(0))
    }

    it("shout get rev_ids and timestamps") {
      val q = ("q" -> "\"* Legal\"") ~ ("fields" -> List("rev_id", "timestamp"))
      val json = retrieve(d.address, q)
      assert(JInt(1) == (json \ "hits_all"))
      assert(JString("233192") ==
        (json \ "hits")(0)(0))
      assert(JString("2001-01-21T02:12:21Z") ==
        (json \ "hits")(0)(1))
    }

    it("should process namespace and page_id -specifying queries") {
      val q = ("q" -> "namespace:0 page_id:12") ~ ("fields" -> "rev_id")
      val json = retrieve(d.address, q)
      assert(JString("+namespace:0 +page_id:12") ==
        (json \ "parsed_query"))
      assert(JInt(1) ==
        (json \ "hits_all"))
      assert(JString("18201") == (json \ "hits")(0)(0))
    }

    val d_s1 = daemon(new SimpleNGramAnalyzer(1), doc)
    val d_h1 = daemon(new HashedNGramAnalyzer(1, 2, 9876), doc)

    def phraseQueries(name:String, address: InetSocketAddress) {
      it(f"should process non-phrase queries (${name})") {
        val q = ("q" -> "subject cover") ~ ("fields" -> "rev_id")
        val json = retrieve(address, q)
        assert(JInt(2) == json \ "hits_all")
      }

      it(f"should process phrase queries (${name})") {
        val q = ("q" -> "\"subject cover\"") ~ ("fields" -> "rev_id")
        val json = retrieve(address, q)
        assert(JInt(1) == json \ "hits_all")
        assert(JString("233192") == (json \ "hits")(0)(0))
      }
    }

    phraseQueries("simple ngram", d_s1.address)
    phraseQueries("hashed ngram", d_h1.address)

  }

  describe("SearcherDaemon with three small documents") {

    val d = n3daemon("""233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'
18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'
12345	10	0	'Accessiblecomputing'	1980043141	u'*'	False	99	u'Automated conversion'	0:1:u'[[fr:ABC]]\\n'
""")

    it("should process OR queries") {
      val q = ("q" -> "rev_id:12345 OR page_id:12") ~ ("fields", "rev_id")
      val json = retrieve(d.address, q)
      assert(JInt(2) == (json \ "hits_all"))
      assert(Set("12345", "18201") ==
        (json \ "hits" \\ classOf[JString]).toSet)
    }

    it("should process '?' queries") {
      val q = ("q" -> "namespace:0 AND NOT \\[\\[?") ~ ("fields", "rev_id")
      val json = retrieve(d.address, q)
      assert(JInt(1) == json \ "hits_all")
      assert(JString("233192") == (json \ "hits")(0)(0))
    }

  }

  describe("SearcherDaemon with another three small documents") {

    val d = n1daemon("""233192	10	0	'Accessiblecomputing'	980043141	u'*'	False	99	u'RoseParks'	0:1:u'This subject covers\\n\\n* AssistiveTechnology\\n\\n* AccessibleSoftware\\n\\n* AccessibleWeb\\n\\n* LegalIssuesInAccessibleComputing\\n\\n'
18201	12	0	'Anarchism'	1014649222	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'
18210	12	0	'Anarchism'	1014749333	u'Automated conversion'	True	None	u'Conversion script'	9230:1:u'[[talk:Anarchism|'	9252:1:u']]'	9260:1:u'[[Anarchism'	9276:1:u'|/Todo]]'	9292:1:u'talk:'	9304:-1:u'/Talk'	9464:1:u'\\n'
""")

    it("should produce collapsed hits by month") {
      val q = ("q" -> "\"/Todo\"") ~ ("fields" -> "rev_id") ~ ("collapse_hits" -> "month")
      val json = retrieve(d.address, q)
      
      assert(JInt(2) == json \ "hits_all")
      assert(JString("2002-02") == (json \ "hits")(0)(0))
      assert(List("18201", "18210") == (json \ "hits")(0)(1) \\ classOf[JString])
    }

    it("should produce collapsed hits by day") {
      val q = ("q" -> "\"/Todo\"") ~ ("fields" -> "rev_id") ~ ("collapse_hits" -> "day")
      val json = retrieve(d.address, q)
      
      assert(JInt(2) == json \ "hits_all")
      assert(JString("2002-02-25") == (json \ "hits")(0)(0))
      assert(JString("2002-02-26") == (json \ "hits")(1)(0))
      assert(JString("18201") == (json \ "hits")(0)(1)(0)(0))
      assert(JString("18210") == (json \ "hits")(1)(1)(0)(0))
      assert(List("18201", "18210") ==
        (for ( JArray(List(JString(_), y)) <- json; JString(z) <- y ) yield z))
    }

    it("should produce collapsed hit counts by day") {
      val q = ("q" -> "\"/Todo\"") ~ ("collapse_hits" -> "day")
      val json = retrieve(d.address, q)
      
      assert(JInt(2) == json \ "hits_all")
      assert(JString("2002-02-25") == (json \ "hits")(0)(0))
      assert(JString("2002-02-26") == (json \ "hits")(1)(0))
      assert(JInt(1) == (json \ "hits")(0)(1))
      assert(JInt(1) == (json \ "hits")(1)(1))
    }

  }

}
